### ISSUE TITLE
Super Resolution Diffusion Model

### ISSUE THREAD

**Issue by dvIdol** [huggingface/diffusers#463]
Hello.

I am very interested in the unconditional image generation pipelines. Like in this example here:
https://github.com/huggingface/diffusers/tree/main/examples/unconditional_image_generation

I have trained a network that is 128x128 and it gives very good results for what I need. However the resolution is very low.

The main diffusers readme mentions a super resolution diffusion model that comes after the low resolution model. How do I make this model? There are no examples and everything seems to be turning to stable diffusion. Is there any guide for how to train a low to high resolution diffusion model?

Thank you for making such a library, it is very good.

**Comment by patrickvonplaten** [comment]
Related to https://github.com/huggingface/diffusers/issues/146

**Comment by anton-l** [comment]
@patil-suraj IIRC you had plans for an SR example too? I might not have bandwidth in the next few weeks, but can work on SR after, if it's not high on your list.

**Comment by patil-suraj** [comment]
Yes, that's on my todo list. But if anyone is interested feel free to open a PR, happy to help :)

**Comment by pcuenca** [comment]
I'm interested in this too, and it's becoming relevant for the on-going fast.ai course. I might have some time to start working on this in a few days and/or help @anton-l and @patil-suraj when they do :)

**Comment by pcuenca** [comment]
My thought exactly :)

**Comment by duongna21** [comment]
@patil-suraj @pcuenca I can spend time implementing SR example this weekend (PyTorch & Flax).

**Comment by patil-suraj** [comment]
Reopening this issue as it's related to training super-res model.

**Comment by patil-suraj** [comment]
Also, thanks to @duongna21 a super resolution model is now available in `diffusers` 

```python
from diffusers import LDMSuperResolutionPipeline
from PIL import Image

pipe = LDMSuperResolutionPipeline.from_pretrained('CompVis/ldm-super-resolution-4x-openimages')
pipe.to('cuda')

img = Image.open('low_resolution.jpg')
super_img = pipe(img, num_inference_steps=100, eta=1)
super_img['images'][0]
```

**Comment by ElliotQi** [comment]
@patil-suraj Hi, How is it going? There's an unofficial repo with much attention: https://github.com/Janspiry/Image-Super-Resolution-via-Iterative-Refinement 

> That's awesome Pedro! I'm looking at implementing SR3 https://iterative-refinement.github.io/ for this task.

**Comment by patil-suraj** [comment]
Haven't really started anything yet, thanks for sharing the repo.

**Comment by basab-gupta** [comment]
I also wanted to share the repo to OpenAI's guided diffusion: [Guided Diffusion](https://github.com/openai/guided-diffusion). SR3 uses the improved version of DDPM as proposed by OpenAI for its experiments. I think you might also find this useful for the implementation of SR3 or even it's follow up model, Palette. Here is a link to the paper that introduced Palette: [Palette: Image-to-Image Diffusion Models](https://arxiv.org/pdf/2111.05826.pdf) and the authors' website [authors' website](https://iterative-refinement.github.io/palette/)

**Comment by patil-suraj** [comment]
@basab-gupta, if you are interested, feel free to start working on it; happy to help with the PR :) 

We can add this example under `examples/research_projects` directory.

**Comment by basab-gupta** [comment]
@patil-suraj Thank you! Do you mean add a link to the guided diffusion repo to  `examples/research_projects`?

I'll try to get started with the implementation. Also, feel free to HMU in case anyone else is interested on working on this together :)

**Comment by patil-suraj** [comment]
I meant to add a training script leveraging `diffusers`.

**Comment by marc-gav** [comment]
I will join you on this script @basab-gupta !

**Comment by basab-gupta** [comment]
Hi! @patil-suraj, we have a small update for you. We managed to set a training script up. However, the loss plateaus out after a point when we run the training. We were thinking of adding a few modifications from [Improved Denoising Diffusion Probabilistic Models](https://arxiv.org/pdf/2102.09672)  and were wondering what you thought of them?
1. Add a cosine scheduler in the DDPM Scheduler 
2. Letting the model learn the variance (this is already in the DDPM Scheduler)

**Comment by basab-gupta** [comment]
Hi @patil-suraj. We have an update for you. We managed to fix our problem with the loss from our previous post. We now have a working implementation of the SR3 model that uses the HF diffusers. Here are some preliminary results from our experiments. [Preliminary Results of 8x super resolution](https://wandb.ai/basabs_master_thesis/sr3_HF_Diffusers/reports/Weave-Table-23-02-16-16-12-26---VmlldzozNTczMjU5?accessToken=k3shbuw9bybnjo9hf6vak1r9hizqwtcur9kgb2utg5mk8tp8nxb5n34va99str8n)

The results however, still do not look quite as good. We are currently working on tuning the hyperparameters to optimize the results and will hopefully get back to you soon with more positive updates :)

**Comment by patrickvonplaten** [comment]
Very cool!

**Comment by basab-gupta** [comment]
@patrickvonplaten Danke :)

**Comment by ElliotQi** [comment]
Hi @basab-gupta @marc-gav ! Thanks for your contribution. There's a problem about sr3: 
As is shown in Fig 12. of the paper, SR3 (Other Google diffusion models) uses noise level sampling during training, it enables the use of different noise schedules during the inference. But I always get noisy output with less testing timestep than training timestep. Did you do some experiments about the different inference timesteps?

**Comment by basab-gupta** [comment]
Hi @ElliotQi! We are still working the inference script to make sure that it allows us to vary the noise schedule and number of steps separately from the ones used in training. Unfortunately, because these models take a while to train, our progress has been a bit slow.

Regarding your question, have you tried adjusting the values of $\beta_{0}, \beta_{N}$, and $N$ during inference? To my understanding, the authors of SR3 fix N (the number of reverse steps) at 100 and then do a hyperparameter sweep to find the best combination of beta values. They use FID scores of their validation dataset to optimize the hyperparameters. We will let you know once we've made some progress on our inference script.

The authors of SR3 use the noise conditioning described in the [Wavegrad paper](https://openreview.net/pdf?id=NsMLjcFaO8O) which is another diffusion model published by the Google Brain team and is used for voice synthesis. I came across this [useful repository](https://github.com/coqui-ai/TTS/blob/2c9f00a808e0aa76a82af2e8b325abb71f50d1df/TTS/bin/tune_wavegrad.py) that has a script to tune the Wavegrad model to find the best inference schedule. Maybe you could take a look at that? Alternatively, I believe you could also use something like Optuna or WanDB to do the hyperparameter tuning for you.

**Comment by ElliotQi** [comment]
@basab-gupta Danke! :)
I tested several values of beta, but no one got good results. I'm still trying some hyperparameters for better performance, thanks for sharing the repo of Wavegrad.
In fact, I noticed that [Deblurring paper](https://openaccess.thecvf.com/content/CVPR2022/papers/Whang_Deblurring_via_Stochastic_Refinement_CVPR_2022_paper.pdf) used this continuous noise schedule to achieve Perception-Distortion trade-off. It's so amazing that I'm tuning my model to reproduce this.
Thanks for your advice~

**Comment by sayakpaul** [comment]
@asomoza WDYT about this?

**Comment by asomoza** [comment]
This shouldn't be an issue anymore, aside from the research aspect, IMO the best results right now are using a GAN SR model in combination with a diffusion model. There's also some good models like [SUPR](https://github.com/Fanghua-Yu/SUPIR) and the [Flux Upscaler Controlnet](https://huggingface.co/jasperai/Flux.1-dev-Controlnet-Upscaler)

If someone still wants to discuss the research aspect of these or how to train them, it would be best to open a discussion or directly ask in the repos of the people that trained them.
